### PR TITLE
Add support for 'reduce_on_plateau' learning rate scheduler

### PR DESCRIPTION
**Pull request by SvenVw**
Hi there! I've created this pull request to enhance the functionality of the package by adding support for the `reduce_on_plateau` learning rate scheduler. Through my experience with other networks, I've observed that this particular scheduler often leads to improved model performance.

Currently, the documentation states that for the `lr_scheduler` parameter in `tabnet_config`, any other `torch::lr_scheduler` function can be passed. However, when I attempted to do so, I encountered an error. The issue arises because during (pre)training, only `scheduler$step()` is supported, while `reduce_on_plateau` requires the loss as input and thus needs `scheduler$step(loss)`.

By submitting this pull request, I aim to introduce this missing functionality to this nice and usefull package 😃. Please let me know if there's anything else I can provide or do to support this pull request.

### CODE REVIEW

**Comment by cregouby** [comment]
Hello @SvenVw 

Thanks for this work ! And very happy to see the package being used !

Currently the 4 tests impacted  in `test_pretraining.R` and  `test-hardhat_parameters.R:97:3` are failing but I don't think it is a big deal to fix them :

It seems to me that the condition in  https://github.com/SvenVw/tabnet/blob/f5cefa41cb7886b314d5555fa2f004663001205f/R/model.R#L545-L551 could be rework to first manage the case `if (config$lr_scheduler == "reduce_on_plateau") {` and use the `else {` for the other lr_shedulers.

Then there is a typo in your 3rd test (see the comment)

Finally, would you add your name as contributor in the author list of "DESCRIPTION" file ? and add a bullet mentioning the new feature in the NEWS.md file ?

Thanks a lot
Christophe

File: tests/testthat/test-pretraining.R [hunk 1]
```diff
@@ -128,6 +128,26 @@ test_that("lr scheduler works", {
...
+test_that("lr scheduler reduce_on_plateau works", {
+
+  expect_error(
+    fit <- tabnet_pretrain(x, y, epochs = 3, lr_scheduler = "reduce_on_plateau",
+                           lr_decay = 0.1, step_size = 1),
+    regexp = NA
+  )
+
+  sc_fn <- function(optimizer) {
+    torch::lr_step(optimizer, step_size = 1, gamma = 0.1)
```
Comment (cregouby) [inline tests/testthat/test-pretraining.R:140]: should be `torch::lr_reduce_on_plateau(optimizer, step_size = 1, gamma = 0.1)`, I think.

File: tests/testthat/test-pretraining.R [hunk 2]
```diff
@@ -128,6 +128,26 @@ test_that("lr scheduler works", {
...
+  }
+
+  expect_error(
+    fit <- tabnet_pretrain(x, y, epochs = 3, lr_scheduler = sc_fn,
+                           lr_decay = 0.1, step_size = 1),
+    regexp = NA
```
Comment (cregouby) [inline tests/testthat/test-pretraining.R:146]: This test is expected to fail as the condition `if (config$lr_scheduler == "reduce_on_plateau")` will not be met here. But that's OK, I think. so I would turn it into `regex = "comparison"` or anything that is in the error message

**Review by cregouby — CHANGES_REQUESTED** [review]

File: tests/testthat/test-pretraining.R (hunk 1 above)
Comment (SvenVw) [reply tests/testthat/test-pretraining.R]: Solved in f80e7bd31ad9313341923ef3c6273d7a22271bb8

**Comment by SvenVw** [comment]
Hi Christophe,

Thanks for your feedback! I have modified the if statement so that it checks if the step function has the argument `metrics`  present or not. In that case it still supports if the user provides `NULL` or a `torch::lr_scheduler` function. Otherwise if you check if `config$lr_scheduler == 'reduce_on_plateau` you get an error if `lr_scheduler` is `NULL`. This implementation seems a little bit cumbersome to me, so feel free to suggest an improvement.

**Comment by cregouby** [comment]
Thanks a lot @SvenVw for this !
